File: examples/word-run-style.ts
/**
 * 例子（Epic-41）：演示 Run.styleId（rStyle 引用 Character Style）。
 *
 * 跑法：
 *   bun run examples/word-run-style.ts <output.docx>
 */

import { Paragraph, Run, Text, WordprocessingDocument } from "../src/word/index.js";

async function main(): Promise<void> {
  const [outputPath] = process.argv.slice(2);
  if (outputPath === undefined) {
    process.stderr.write("usage: word-run-style <output.docx>\n");
    process.exit(2);
  }

  const doc = WordprocessingDocument.create();
  const body = doc.mainDocumentPart!.document.firstChild()!;
  const append = (e: unknown): void =>
    (body as { appendChild: (e: unknown) => void }).appendChild(e);

  function makeRun(styleId: string | undefined, text: string): Run {
    const r = new Run();
    const t = new Text();
    t.text = text;
    r.appendChild(t);
    if (styleId !== undefined) r.styleId = styleId;
    return r;
  }

  function addLine(...runs: Run[]): void {
    const p = new Paragraph();
    for (const r of runs) p.appendChild(r);
    append(p);
  }

  addLine(
    makeRun(undefined, "Body text with a "),
    makeRun("Strong", "Strong word"),
    makeRun(undefined, " and an "),
    makeRun("Emphasis", "emphasized phrase"),
    makeRun(undefined, "."),
  );
  addLine(makeRun("Hyperlink", "Hyperlink-styled run."));
  addLine(makeRun("Subtitle", "Subtitle style."));

  await doc.saveAsAsync(outputPath);
  process.stdout.write(`Wrote ${outputPath} (mixed Character Style refs)\n`);
}

main().catch((err) => {
  process.stderr.write(`failed: ${(err as Error).stack ?? err}\n`);
  process.exit(1);
});

File: tools/cross-sdk-verify/replicas/WordRunStyle.cs
// Replica of examples/word-run-style.ts
// Creates a docx with 3 paragraphs demonstrating run character style references.

using DocumentFormat.OpenXml;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Wordprocessing;

namespace CrossSdkVerify.Replicas;

public static class WordRunStyle
{
    public static void Run(string outputPath)
    {
        using var doc = WordprocessingDocument.Create(outputPath,
            DocumentFormat.OpenXml.WordprocessingDocumentType.Document);

        var mainPart = doc.AddMainDocumentPart();
        mainPart.Document = new Document();
        var body = new Body();
        mainPart.Document.Append(body);

        // Line 1: mixed run styles
        var p1 = new Paragraph();
        p1.Append(MakeRun(null, "Body text with a "));
        p1.Append(MakeRun("Strong", "Strong word"));
        p1.Append(MakeRun(null, " and an "));
        p1.Append(MakeRun("Emphasis", "emphasized phrase"));
        p1.Append(MakeRun(null, "."));
        body.Append(p1);

        // Line 2
        var p2 = new Paragraph();
        p2.Append(MakeRun("Hyperlink", "Hyperlink-styled run."));
        body.Append(p2);

        // Line 3
        var p3 = new Paragraph();
        p3.Append(MakeRun("Subtitle", "Subtitle style."));
        body.Append(p3);

        body.Append(new SectionProperties());
        mainPart.Document.Save();
    }

    private static Run MakeRun(string? styleId, string text)
    {
        var run = new Run(new Text(text) { Space = SpaceProcessingModeValues.Preserve });
        if (styleId != null)
        {
            var rPr = new RunProperties(new RunStyle { Val = styleId });
            run.PrependChild(rPr);
        }
        return run;
    }
}
